### PR TITLE
Balanced Distribution, Shared Agents support

### PR DESCRIPTION
**Pull request by deepwired**
https://hello-haptik.atlassian.net/browse/AGCT-274
https://hello-haptik.atlassian.net/browse/AGCT-1086

### CODE REVIEW

File: docs/agent-chat/chat-assignment.md [hunk 1]
```diff
@@ -36,9 +36,13 @@ This accounts for Agent Concurrency, Empty Slots and Idle Time. Chat gets assign
 
 3. Chats that are in pending state gets completed when last agent of a team logouts.
 
-4. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+4. If an agent is part of two teams, both following Balanced Distribution, the concurrency of that agent will be doubled.
+i.e, If agent concurrency is C, and is a part of N teams following Balanced distribution, he concurrency will be C times N.
+We recommend keeping a lower agent concurrency for shared agents.
```
Comment (deepwired) [inline docs/agent-chat/chat-assignment.md:41]: Note

Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:41]: My Chats should show the real correct values for number of chats assigned/pinned as a test case

Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:41]: he = his conc..

Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:41]: If agent concurrency is C, and is a part of N teams following Balanced distribution, he concurrency will be C times N.    ---->>> he can get these many chats actually assigned.

File: docs/agent-chat/chat-assignment.md [hunk 2]
```diff
@@ -36,9 +36,13 @@ This accounts for Agent Concurrency, Empty Slots and Idle Time. Chat gets assign
...
 
-4. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+4. If an agent is part of two teams, both following Balanced Distribution, the concurrency of that agent will be doubled.
+i.e, If agent concurrency is C, and is a part of N teams that are following Balanced distribution algorithm, his/her concurrency will be C times N.
+> Note: We recommend keeping a lower agent concurrency for shared agents.
 
-5. Con(s): Does not account for faster agent response times but only the first response times. Faster high performing agents may not be best utilized. 
+5. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+
+6. Con(s): Does not account for faster agent response times but only the first response times. Faster high performing agents may not be best utilized.
```
Comment (1a8e) [inline docs/agent-chat/chat-assignment.md:45]: - Shouldn't this be a continuation of point 5?
 - `Faster high performing agents may not be best utilized.`: seems a bit confusing, not sure what 'faster', 'high performing' mean in this context.

File: docs/agent-chat/chat-assignment.md [hunk 3]
```diff
@@ -36,9 +36,13 @@ This accounts for Agent Concurrency, Empty Slots and Idle Time. Chat gets assign
 
 3. Chats that are in pending state gets completed when last agent of a team logouts.
 
-4. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+4. If an agent is part of two teams, both following Balanced Distribution, the concurrency of that agent will be doubled.
+i.e, If agent concurrency is C, and is a part of N teams that are following Balanced distribution algorithm, his/her concurrency will be C times N.
+> Note: We recommend keeping a lower agent concurrency for shared agents.
```
Comment (1a8e) [inline docs/agent-chat/chat-assignment.md:41]: Should probably add a note stating you can't have agents shared across distribution mode types.

File: docs/agent-chat/chat-assignment.md [hunk 4]
```diff
@@ -36,9 +36,13 @@ This accounts for Agent Concurrency, Empty Slots and Idle Time. Chat gets assign
 
 3. Chats that are in pending state gets completed when last agent of a team logouts.
 
-4. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+4. If an agent is part of two teams, both following Balanced Distribution, the concurrency of that agent will be doubled.
+i.e, If agent concurrency is C, and is a part of N teams that are following Balanced distribution algorithm, his/her concurrency will be C times N.
```
Comment (1a8e) [inline docs/agent-chat/chat-assignment.md:40]:
```suggestion
i.e, If agent's concurrency is C, and they are a part of N teams that are following Balanced distribution algorithm, their effective concurrency will be C times N.
```

**Review by 1a8e — CHANGES_REQUESTED** [review]

File: docs/agent-chat/chat-assignment.md (hunk 3 above)
Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:41]: i'll put this point somewhere higher

File: docs/agent-chat/chat-assignment.md (hunk 2 above)
Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:45]: If you have a single client and multiple businesses set up, all businesses must be the same algorithm

--- need to update this for clients with multiple teams and diff algos

File: docs/agent-chat/chat-assignment.md (hunk 3 above)
Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:41]: mentioned at the bottom, under Notes

File: docs/agent-chat/chat-assignment.md (hunk 2 above)
Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:45]: changed the wordings by a bit

File: docs/agent-chat/chat-assignment.md [hunk 5]
```diff
@@ -36,9 +36,13 @@ This accounts for Agent Concurrency, Empty Slots and Idle Time. Chat gets assign
...
 
-4. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+4. If an agent is part of two teams, both following Balanced Distribution, the concurrency of that agent will be doubled.
+i.e, If an agent's set concurrency on the "Teams" UI is C, and the agent is part of N teams that are following Balanced distribution algorithm, this agent's effective concurrency will be C times N.
+> Note: We recommend keeping a lower agent concurrency for shared agents.
 
-5. Con(s): Does not account for faster agent response times but only the first response times. Faster high performing agents may not be best utilized. 
+5. Optimized for: Chat load as it’s almost evenly distributed across online agents at the time.
+
+6. Con(s): Unlike Speed Resolution, Quick high performing agents may not be best utilized.
```
Comment (medhaanand) [inline docs/agent-chat/chat-assignment.md:45]:
```suggestion
6. Con(s): Unlike Speed Resolution, high performing agents may not be best utilised. 
```

File: docs/agent-chat/chat-assignment.md [hunk 6]
```diff
@@ -59,6 +63,6 @@ Fastest agents get more chats. Every time an agent closes a chat or intermittent
 
 ### Note
 
-1. If you have a single client and multiple businesses set up, all businesses must be the same algorithm
-
-2. If there are multiple clients, one can ideally have different algorithms for all, provided there are no shared agents between the teams of different clients.
+1. If you have a single client and multiple businesses set up, all businesses must use the same algorithm.
+2. If there are multiple clients, one can select different algorithms for each.
```
Comment (medhaanand) [inline docs/agent-chat/chat-assignment.md:67]: This sentence does not need to be present as a separate point here, I think @deepwired

Comment (deepwired) [reply docs/agent-chat/chat-assignment.md:67]: 1 - > single client multi business scenarios
2 - > multi client scenario
3 - > agent level stuff

we should keep 3 acc to me, open to re-wording.

Comment (medhaanand) [reply docs/agent-chat/chat-assignment.md:67]: Alright, rewording it slightly then.

File: docs/agent-chat/chat-assignment.md [hunk 7]
```diff
@@ -59,6 +63,6 @@ Fastest agents get more chats. Every time an agent closes a chat or intermittent
 
 ### Note
 
-1. If you have a single client and multiple businesses set up, all businesses must be the same algorithm
-
-2. If there are multiple clients, one can ideally have different algorithms for all, provided there are no shared agents between the teams of different clients.
+1. If you have a single client and multiple businesses set up, all businesses must use the same algorithm.
```
Comment (medhaanand) [inline docs/agent-chat/chat-assignment.md:66]:
```suggestion
1. If there is a single client and multiple businesses set up, all businesses must use the same algorithm.
```

File: docs/agent-chat/chat-assignment.md (hunk 6 above)
Comment (medhaanand) [inline docs/agent-chat/chat-assignment.md:67]:
```suggestion
2. If there are multiple clients, no compulsion of using same algorithm. One can select different algorithms for each.
```

**Review by medhaanand — APPROVED** [review]

**Review by 1a8e — APPROVED** [review]

**Pull request merged**